ext/phar/phar_object.c

```diff
@@ -4417,19 +4417,40 @@ PHP_METHOD(Phar, extractTo)
 		phar_archive_data *phar;
 all_files:
 		phar = phar_obj->archive;
-		/* Extract all files */
+		/* Extract all files, or if we have jumped here from a potential
+		   directory match, just those manifest entries matching the dir */
 		if (!zend_hash_num_elements(&(phar->manifest))) {
 			RETURN_TRUE;
 		}
 
+		extracted = 0;
 		ZEND_HASH_FOREACH_PTR(&phar->manifest, entry) {
+			/* handle earlier possible directory, see if this leaf entry matches the leading directory */
+			if (NULL != zval_file) {
+				filename = ZSTR_VAL(Z_STR_P(zval_file));
+			}
+			if (NULL != filename) {
+			    if (strncmp(filename, entry->filename, strlen(filename)) || '/' != entry->filename[strlen(filename)]) {
```

Comment:
If I'm reading this correctly, this requires you to pass `dir` and does not allow `dir/`. Is that intentional / consistent with other functionality? Also, is it necessary to account for `\\` or is this normalized in phars?

Reply:
Your read is correct, and I did implement it that way intentionally - under the expectation I'd also update the documentation to call out this requirement and would plan to fix it in a later round of refactoring / improvements to Phar.

I do _feel_ it's consistent. For example, if the manifest contains a file `foo` at the top-level, it can _only_ be accessed by `foo`: `/foo` does not work. Likewise, if a manifest contains `bar/foo`, then only `bar/foo` works: `/bar/foo`, `bar//foo`, and other conventions one might think _should_ work from experience with shell path behavior does not apply.

Happy to address it, but I feel allowing variability here would be inconsistent with the specific path requirements of files.

Reply:
Sounds reasonable, though I'm still a bit leery about this, especially as the original bug report specifically makes use of the trailing slash.

Maybe @remicollet as the maintainer of the zip extension can chime in? It also has a `ZipArchive::extractTo()` method, which also does not seem to support directories, so I'd like to have your opinion on both the change in general and this specific question.